book/src/developers/designs/messaging.md

```diff
@@ -0,0 +1,90 @@
+removed in this situation (again, meaning we can't just refer the message object to an account which has the mail
+addresses)
+
+### Internationalisation
+
+Many of these messages will need to be internationalised in future, so we should avoid embedding text directly in our
+work, and only indicate or store "events" that can be translated to some kind of text based on user preference (like
+langugae).
+
+### Multiple Kani Servers
+
+Kanidm often has mulitple servers, so we need a way to ensure that messages are sent "at least once". But at the same
+time we don't want to send the message _too many_ times, as this can annoy users.
+
+### Different Messaging Requirements
+
+Email can be a right pain, so we need to expect that users may have weird or niche email requirements when they send
+messages. This means we need to try to make this somewhat extensible and detached from our server.
+
+## Requirements
+
+From this set of use cases we can derive a number of basic requirements.
+
+- When we generate a message notification we need to copy the email address(es) we will send to, not relying on the
+  users account details later.
+- The time in which we send a message may not be "now" but delayed, so we need to ensure that we can send messages in
+  the future
+- Messages may need to be sent to multiple email addresses at the same time
+- The text of the message should not be fixed when the event is generated, but the languages of interest to the user
+  should be present, and may have multiple languages.
+- We need a way to queue messages, but also indicate when the message has been dispatched by a provider.
+
+## Design
+
+Rather than embed messaging (email) capability into Kanidm directly, we should create an api where an external messaging
+tool can consume database records from Kanidm when we wish to communicate with a user.
+
+The database records that we generate should include:
+
+- A generic enum indicating the type of message we want to send
+- Fields that may be related to the enum of the message (such as the credential reset token)
+- A datetime when the message should be sent _after_ (aka not-before)
+- The set of mail addresses that the message should be sent to
+- A datetime that indicates when the message was sent
+
+This allows the external tool to do a number of things.
+
+First, messages being an enum allows the external tool to template it's own structure of message, and customise text
+etc. We can also extend this to language preferences in future too.
+
+Second, having the datetime for sending after means that we can delay messages depending on context.
+
+Third we copy the addresses to the message, so that even if the users email has changed, we send it to the users emails
+that existed _at the time_ of message generation.
+
+Finally, by storing when the message was sent we can allow the tool to be "somewhat" HA so that messages won't be sent
```

Comment:
HA ?

Reply:
Highly Available. As in multiple servers that are active.